Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM_2nd.cpp

```diff
@@ -0,0 +1,540 @@
+              Real Hy_xface = MacroscopicProperties::getH_Maxwell(i, j, k, 1, amrex::IntVect(0,1,0), amrex::IntVect(1,0,0), By, M_prev_xface);
+              Real Hz_xface = MacroscopicProperties::getH_Maxwell(i, j, k, 2, amrex::IntVect(0,0,1), amrex::IntVect(1,0,0), Bz, M_prev_xface);
+              // H_bias
+              Real Hx_bias_xface = MacroscopicProperties::face_avg_to_face(i, j, k, 0, amrex::IntVect(1,0,0), amrex::IntVect(1,0,0), Hx_bias);
+              Real Hy_bias_xface = MacroscopicProperties::face_avg_to_face(i, j, k, 0, amrex::IntVect(0,1,0), amrex::IntVect(1,0,0), Hy_bias);
+              Real Hz_bias_xface = MacroscopicProperties::face_avg_to_face(i, j, k, 0, amrex::IntVect(0,0,1), amrex::IntVect(1,0,0), Hz_bias);
+              // H_eff = H_maxwell + H_bias + H_exchange + H_anisotropy ... (only the first two terms are considered here)
+              Real Hx_eff = Hx_bias_xface;
+              Real Hy_eff = Hy_bias_xface;
+              Real Hz_eff = Hz_bias_xface;
+
+              // Real Hx_eff = Hx_xface + Hx_bias_xface;
+              // Real Hy_eff = Hy_xface + Hy_bias_xface;
+              // Real Hz_eff = Hz_xface + Hz_bias_xface;
+
+              // magnetic material properties mag_alpha and mag_Ms are defined at cell nodes
+              // keep the interpolation
+              Real mag_gamma_interp = MacroscopicProperties::macro_avg_to_face(i,j,k,amrex::IntVect(1,0,0),mag_gamma_arr);
+
+              Real a_temp_dynamic_coeff = PhysConst::mu0 * std::abs(mag_gamma_interp) *
+                        (1.0 + std::pow(MacroscopicProperties::macro_avg_to_face(i,j,k,amrex::IntVect(1,0,0),mag_alpha_arr), 2.0))/ 2.0;
+
+          // calculate a_temp_xface
+              // x component on x-faces of grid
+              a_temp_xface(i, j, k, 0) = -( dt * a_temp_dynamic_coeff * Hx_eff + a_temp_static_xface(i, j, k, 0) );
+
+              // y component on x-faces of grid
+              a_temp_xface(i, j, k, 1) = -( dt * a_temp_dynamic_coeff * Hy_eff + a_temp_static_xface(i, j, k, 1) );
+
+              // z component on x-faces of grid
+              a_temp_xface(i, j, k, 2) = -( dt * a_temp_dynamic_coeff * Hz_eff + a_temp_static_xface(i, j, k, 2) );
+
+              // calculate M_xface
+              // x component on x-faces of grid
+              M_xface(i, j, k, 0) = MacroscopicProperties::getM_half (i, j, k, 0, a_temp_xface, b_temp_xface);
+
+              // y component on x-faces of grid
+              M_xface(i, j, k, 1) = MacroscopicProperties::getM_half (i, j, k, 1, a_temp_xface, b_temp_xface);
+
+              // z component on x-faces of grid
+              M_xface(i, j, k, 2) = MacroscopicProperties::getM_half (i, j, k, 2, a_temp_xface, b_temp_xface);
+
+              // x component on x-faces of grid
+              M_error_xface(i, j, k, 0) = std::abs((M_xface(i, j, k, 0) - M_prev_xface(i, j, k, 0))) / (std::abs(M_prev_xface(i, j, k, 0)) + 1.0e-12);
+
```

Comment:
why adding `1e-12` here?

Reply:
Probably to prevent divide by zero?
However things brings up a different issue: you are looking at relative error, not absolute error.  For problems where M varies throughout the domain from a huge value to very small values, a better design may be to look at the change relative to the largest value of M in the domain?  Or something less susceptible to numerical issues such as dividing by very small numbers.

Reply:
agreed...

Reply:
I choose to look at the change relative to the largest value of each M component in the domain. Check the codes to see if you have more suggestions.